File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int max = Integer.MIN_VALUE;

    int left = 0;
    int right = heights.length - 1;

    while(left < right){
        max = Math.max(max,(right - left)*Math.min(heights[left],heights[right]));
        if(heights[left] > heights[right]) right--;
        else if(heights[left] < heights[right])  left++;
        else {
            right--;
            left++;
        }
    }
    
    return max;        
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:
    int characterReplacement(std::string s, int k) {
        std::unordered_map<char, int> count;
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); ++right) {
            count[s[right]]++;

            int maxFreq = 0;
            for (const auto& pair : count) {
                maxFreq = std::max(maxFreq, pair.second);
            }

            while ((right - left + 1) - maxFreq > k) {
                count[s[left]]--;
                left++;

                maxFreq = 0;
                for (const auto& pair : count) {
                    maxFreq = std::max(maxFreq, pair.second);
                }
            }

            maxLen = std::max(maxLen, right - left + 1);
        }

        return maxLen;
}
};
